charging_station.py

```diff
@@ -0,0 +1,156 @@
+            new_stock["fast"][1]=self.battery_stock["fast"][time][1] + (self.efficiency*max(0,load_battery["fast"][1])+min(0,load_battery["fast"][1])/self.efficiency)*self.dt
+            # We update the new stock of each batteries "fast"
+
+
+        for speed in ["slow","fast"] :
+            for i in range(2):
+                if new_stock[speed][i] < 0:
+                    load_battery[speed][i] = -(self.battery_stock[speed][time][i])/(self.efficiency*self.dt)
+                    new_stock[speed][i] = 0
+            # We can't discharge the batterie under 0
+
+                elif new_stock[speed][i] > c_max[speed][i]:
+                    load_battery[speed][i] = (c_max[speed][i] - self.battery_stock[speed][time][i] ) / (self.efficiency*self.dt)
+                    new_stock[speed][i] = c_max[speed][i]
+            # We can't charge the batteries over their maximum capacities
+
+        for speed in ["slow","fast"] :
+            for i in range(2):
+                if abs(load_battery[speed][i]) >= p_max[speed][i]:
+                    load_battery[speed][i] = p_max[speed][i]*np.sign(load_battery[speed][i])
+            # Can't put more power than p_max
+
+        for speed in ["slow","fast"] :
+            for i in range(2):
+                self.battery_stock[speed][time+1][i]=new_stock[speed][i]
+            # Update of batteries stocks
+                if time == self.arrival[speed][i]-1:
+                    self.battery_stock[speed][time+1][i] = self.battery_stock[speed][int(self.depart[speed][i])][i]-4
+                    # When the cars comes back it has lost 4 kWh in the battery
+                self.load_battery_periode[speed][time][i] = load_battery[speed][i]
+
+        return load_battery # We return load_battery clear of the player's potential mistakes
+
+
+
+
+    def nb_cars(self,time):
+        s = 0
+        for i in range(self.nb_slow_max):
+            if (self.depart["slow"][i]<time) and (self.arrival["slow"][i]>time):
+                self.here["slow"][i]=0
+            else:
+                self.here["slow"][i]=1
+                s+=1
+        f = 0
+        for j in range(self.nb_fast_max):
+            if (self.depart["fast"][j]<time) and (self.arrival["fast"][j]>time):
+                self.here["fast"][j]=0
+            else:
+                self.here["fast"][j]=1
+                f+=1
+        self.nb_slow = s
+        self.nb_fast = f
+        # Acctualise how many cars and which are at the station at t = time.
+
+
+    def penalty(self,time):
+        for speed in ["slow","fast"] :
+            for i in range(2):
+                if time == self.depart[speed][i] and self.battery_stock[speed][time][i]/40 < 25:
+                    self.bill[time]+=5
+        # If at the departure time of the veicule its battery isn't charged at least at 25% then you pay a 5e fine
+
+
+    def take_decision(self, time):
```

Comment:
It is better if this function take into account the presence of EV, imposing a zero load when EV is not present

Reply:
Donc si je comprends bien j'intègre dans la fonction take decision un test qui met load battry a 0 si la voiture n'est pas là ? Je ne penses pas que ce soit nécessaire car la fonction update_battry stock corrige le battery load du jouer si la voiture n'est pas là. Mais l'intégrer ne coûte vraiment pas grand chose donc c'est vous qui me dites.

Reply:
Oui c'est très bien